sdk/identity/identity/src/credentials/azurePowerShellCredential.ts

```diff
@@ -55,83 +58,77 @@ export const powerShellErrors = {
-}
+const isNotInstalledError = (err: Error) => err.message.match(powerShellErrors.installed);
 
 /**
- * This credential will use the currently logged-in user login information via the Azure Power Shell command line tool.
- * To do so, it will read the user access token and expire time with Azure Power Shell command `Get-AzAccessToken -ResourceUrl {ResourceScope}`.
- * To be able to use this credential, ensure that:
- * - Install the Azure Az PowerShell module with: `Install-Module -Name Az -Scope CurrentUser -Repository PSGallery -Force`.
- * - You have already logged in via the 'az' tool using the command "Connect-AzAccount" from the command line.
+ * This credential will use the currently logged-in user information from the
+ * Azure PowerShell module. To do so, it will read the user access token and
+ * expire time with Azure PowerShell command `Get-AzAccessToken -ResourceUrl {ResourceScope}`
+ *
+ * To be able to use this credential:
+ * - Install the Azure Az PowerShell module with:
+ *   `Install-Module -Name Az -Scope CurrentUser -Repository PSGallery -Force`.
+ * - You have already logged in to Azure PowerShell using the command
+ * `Connect-AzAccount` from the command line.
  */
 export class AzurePowerShellCredential implements TokenCredential {
-  private readonly powerShellCommand: string;
+  private commandStack: string[] = [formatCommand("pwsh")];
 
   /**
-   * Creates an instance of the ClientCertificateCredential with the details
-   * needed to authenticate against Azure Active Directory with a certificate.
-   *
-   * @param AzurePowerShellCredentialOptions - Optional parameters. For now accepts a `useLegacyPowerShell` property to indicate if legacy powershell should be used for authentication.
+   * Creates an AzurePowerShellCredential, which will authenticate against
+   * Azure Active Directory using an existing login session created using
+   * `Connect-AzAccount` in PowerShell.
    */
-  constructor(options?: AzurePowerShellCredentialOptions) {
-    if (options?.useLegacyPowerShell) {
-      this.powerShellCommand = formatCommand("powershell");
-    } else {
-      this.powerShellCommand = formatCommand("pwsh");
+  constructor() {
+    if (isWindows) {
+      this.commandStack.push(formatCommand("powershell"));
     }
   }
+
   /**
-   * Gets the access token from Azure Power Shell
+   * Gets the access token from Azure PowerShell
    * @param resource - The resource to use when getting the token
    */
   private async getAzurePowerShellAccessToken(
     resource: string
   ): Promise<{ Token: string; ExpiresOn: string }> {
-    try {
-      await runCommands([[this.powerShellCommand, "/?"]]);
-    } catch (e) {
-      throw new Error(
-        `Unable to execute "${this.powerShellCommand}". Ensure that it is installed in your system.`
-      );
-    }
+    for (const powerShellCommand of this.commandStack) {
+      try {
+        await runCommands([[powerShellCommand, "/?"]]);
+      } catch (e) {
+        continue;
```

Comment:
should this pop `this.commandStack` so if we end up falling back to powershell.exe we only try pwsh.exe the first time?